### PR TITLE
feature(batch-action) add reprocessing endpoint

### PR DESCRIPTION
**Pull request by rprzedzik**
# Description

Add reprocessing endpoint

Issue #1556
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [ ] Unit Tests
- [x] Behat Test

# Checklist:
- [x] I have read the contribution requirements and fulfill them.
- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: module/batch-action/src/Application/Controller/Api/ReprocessBatchAction.php [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+     * @SWG\Response(
+     *     response=200,
+     *     description="Returns batch information",
+     * )
+     *
+     * @ParamConverter(class="Ergonode\BatchAction\Domain\Entity\BatchAction", name="action")
+     */
+    public function __invoke(BatchAction $action, Request $request): void
+    {
+        $type = $action->getType()->getValue();
```
Comment (piotrkreft) [inline module/batch-action/src/Application/Controller/Api/ReprocessBatchAction.php:75]: If BatchAction is not in `WAITING_FOR_DECISION` status an exception should be thrown

File: module/batch-action/src/Application/Controller/Api/ReprocessBatchAction.php [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+     *     default="en_GB"
+     * )
+     * @SWG\Parameter(
+     *     name="action",
+     *     in="path",
+     *     type="string",
+     *     description="Batch action id",
+     * )
+     * @SWG\Response(
+     *     response=200,
```
Comment (piotrkreft) [inline module/batch-action/src/Application/Controller/Api/ReprocessBatchAction.php:67]:
```suggestion
     *     response=204,
```

File: module/batch-action/src/Application/Controller/Api/ReprocessBatchAction.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+     *     description="Returns batch information",
```
Comment (piotrkreft) [inline module/batch-action/src/Application/Controller/Api/ReprocessBatchAction.php:68]: Does not return information

File: module/batch-action/src/Application/Form/BatchActionCountForm.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                        TextType::class,
+                    );
+                } else {
+                    $form->add(
+                        'filter',
+                        BatchActionFilterType::class,
+                    );
+                }
+            })
+            ->add(
```
Comment (piotrkreft) [inline module/batch-action/src/Application/Form/BatchActionCountForm.php:44]: Seem redundant

File: module/batch-action/src/Application/Form/BatchActionForm.php [hunk 5]
```diff
@@ -8,17 +8,10 @@
...
-use Symfony\Component\Form\AbstractType;
 use Symfony\Component\Form\Extension\Core\Type\TextType;
 use Symfony\Component\Form\FormBuilderInterface;
-use Symfony\Component\Form\FormEvent;
-use Symfony\Component\Form\FormEvents;
-use Symfony\Component\OptionsResolver\OptionsResolver;
-use Symfony\Component\Form\Extension\Core\Type\CheckboxType;
 
-class BatchActionForm extends AbstractType implements BatchActionFormInterface
+class BatchActionForm extends AbstractBatchActionForm
```
Comment (piotrkreft) [inline module/batch-action/src/Application/Form/BatchActionForm.php:14]: If possible I'd remove default forms.

File: module/batch-action/src/Application/Form/BatchActionReprocessForm.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Ergonode\BatchAction\Application\Form;
+
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormBuilderInterface;
+
+class BatchActionReprocessForm extends AbstractBatchActionReprocessForm
```
Comment (piotrkreft) [inline module/batch-action/src/Application/Form/BatchActionReprocessForm.php:14]: If possible I'd remove default form

File: module/batch-action/src/Application/Form/Model/BatchActionCountFormModel.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+     * )
+     */
+    public ?string $type = null;
+
+    /**
+     * @Assert\Valid()
+     * @Assert\NotBlank()
+     * @AllFilterDisabled()
+     *
+     * @var string|BatchActionFilterFormModel $filter
```
Comment (piotrkreft) [inline module/batch-action/src/Application/Form/Model/BatchActionCountFormModel.php:30]: Seems like `use` is missing

File: module/batch-action/src/Application/Form/ReprocessingBatchActionForm.php [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+use Ergonode\BatchAction\Application\Form\Type\BatchActionFilterType;
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\Form\Extension\Core\Type\CheckboxType;
+
+class ReprocessingBatchActionForm extends AbstractType implements BatchActionFormInterface
```
Comment (piotrkreft) [inline module/batch-action/src/Application/Form/ReprocessingBatchActionForm.php:21]: It seems redundant

File: module/batch-action/src/Domain/Command/ReprocessBatchActionCommand.php [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+    private bool $autoEndOnErrors;
+
+    /**
+     * @param mixed $payload
+     */
+    public function __construct(
+        BatchActionId $id,
+        $payload = null,
+        bool $autoEndOnErrors = true
```
Comment (piotrkreft) [inline module/batch-action/src/Domain/Command/ReprocessBatchActionCommand.php:26]: Something off here - value can be nullable if not passed - then should be taken from the current action - this is not done in controller so I assume there's an error somewhere.

File: module/batch-action/src/Infrastructure/Handler/ReprocessBatchActionCommandHandler.php [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        $this->repository = $repository;
+    }
+
+    public function __invoke(ReprocessBatchActionCommand $command): void
+    {
+        $batchAction = $this->repository->load($command->getId());
+        Assert::isInstanceOf($batchAction, BatchAction::class);
+        $batchAction->setPayload($command->getPayload());
+        $batchAction->setStatus(new BatchActionStatus(BatchActionStatus::PROCESSED));
```
Comment (piotrkreft) [inline module/batch-action/src/Infrastructure/Handler/ReprocessBatchActionCommandHandler.php:31]:
```suggestion
        $batchAction->setStatus(new BatchActionStatus(BatchActionStatus::PRECESSED));
```
it seems like a typo in status name - I think we do not have PROCESSED one

**Review by piotrkreft — CHANGES_REQUESTED** [review]

File: module/batch-action/src/Application/Form/Model/BatchActionCountFormModel.php (hunk 7 above)
Comment (rprzedzik) [reply module/batch-action/src/Application/Form/Model/BatchActionCountFormModel.php:30]: same namspace

**Review by wfajczyk — DISMISSED** [review]

File: module/batch-action/src/Application/Form/BatchActionReprocessingFormInterface.php [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+/**
+ * Copyright © Bold Brand Commerce Sp. z o.o. All rights reserved.
+ * See LICENSE.txt for license details.
+ */
+
+declare(strict_types=1);
+
+namespace Ergonode\BatchAction\Application\Form;
+
+interface BatchActionReprocessingFormInterface
```
Comment (piotrkreft) [inline module/batch-action/src/Application/Form/BatchActionReprocessingFormInterface.php:11]:
```suggestion
interface BatchActionReprocessFormInterface
```
More in line with abstraction and other classes

**Review by piotrkreft — CHANGES_REQUESTED** [review]

**Review by piotrkreft — APPROVED** [review]

**Review by wfajczyk — APPROVED** [review]

**Pull request merged**